bootloader/grub.go

```diff
@@ -532,7 +541,10 @@ func (g *grub) getGrubRecoveryModeTrustedAssets() ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
-	return []string{assets.shimBinary, assets.grubBinary}, nil
+	if osutil.FileExists(filepath.Join(g.rootdir, assets.fallbackBinary)) {
```

Comment:
what happens during installation or refresh with this condition? does it change from false to true at a reasonable point? should it based on what is in the rootdir or what is in the gadget? should that vary if the bootloader was produced via Find vs ForGadget?

Reply:
In `doInstall` the assets are updated before the resealing happens:
https://github.com/snapcore/snapd/blob/a90520700167916c204872c93ec74a14c517af86/overlord/snapstate/snapstate.go#L548-L563
("update-gadget-assets" before "update-gadget-cmdline") so in principle the file should be there already before resealing.

Reply:
I think the confusing thing is that it is already broken. Because we call `NewBootFile("", ta, RoleRecovery)` and `NewBootFile("", ta, RoleRunMode)`, we probably do not do a proper first reseal (the one with all bootchains).

We do pass the kernel path to `BootChain` and `RecoveryBootChain`. But not the gadget. So we are missing some information here.

Reply:
sorry, what is exactly already broken? if the first sealing was broken we would not be able to boot the first time?  at installation time the gadget assets reach the seal code via the TrustedAssetsInstallObserver that is then passed to  MakeRunnableSystem but you are probably referring to something different but I'm not sure what

Reply:
Sorry, I was confused. I forgot that for bootloader assets, the BootFile is not passed directly to the boot chain we pass to secboot. We only use the path as index to find the real file in CurrentTrustedBootAssets in the modeenv.

Reply:
@pedronis I have PR #13398 with a fix for that.

It seems I cannot push to this PR. Should I just open that other PR to continue there?

Reply:
It is commit https://github.com/snapcore/snapd/pull/13398/commits/cb88e793ac6338859ff5b38796d584b6b6bdbffc

Reply:
@pedronis I have opened #13412. Maybe you can have a quick pass to see if the approach is reasonable.